### PR TITLE
🐛 + ✨ [Button] Bug fixes, small improvements, new "select" disclosure

### PR DESCRIPTION
**Pull request by beefchimi**
> **Tophat:** I recommend using the provided `Playground` snippet to more easily tophat each combination of `<Button />`

<details>
<summary><strong>Playground code:</strong></summary>

```tsx
import React, {useCallback, useState} from 'react';
import {OnlineStoreMajor} from '@shopify/polaris-icons';

import {Button, Card, ChoiceList, Layout, Page} from '../src';

const childrenOptions = {
  none: undefined,
  short: 'Short text',
  long: 'This long line of text will wrap unless truncate is true',
};
const sizeOptions = {
  none: undefined,
  slim: 'slim',
  medium: 'medium',
  large: 'large',
};
const textAlignOptions = {
  none: undefined,
  left: 'left',
  center: 'center',
  right: 'right',
};
const disclosureOptions = {
  none: undefined,
  up: 'up',
  down: 'down',
  select: 'select',
};
const toggleOptions = [
  'primary',
  'destructive',
  'outline',
  'fullWidth',
  'plain',
  'monochrome',
  'icon',
  'stretchContent',
  'connectedDisclosure',
  'url',
  'external',
  'download',
  'submit',
  'disabled',
  'loading',
  'pressed',
  'role',
  'ariaExpanded',
];

const mockConnectedDisclosure = {
  actions: [
    {
      content: 'Save and mark as ordered',
    },
  ],
};
const mockRole = 'menuitem';
const mockUrl = 'https://shopify.ca';

function mockEvent(label = '') {
  // eslint-disable-next-line no-console
  return () => console.log(label);
}

export function Playground() {
  const [children, setChildren] = useState(['none']);
  const handleChildrenChange = useCallback((value) => setChildren(value), []);

  const [size, setSize] = useState(['none']);
  const handleSizeChange = useCallback((value) => setSize(value), []);

  const [textAlign, setTextAlign] = useState(['none']);
  const handleTextAlignChange = useCallback((value) => setTextAlign(value), []);

  const [disclosure, setDisclosure] = useState(['none']);
  const handleDisclosureChange = useCallback(
    (value) => setDisclosure(value),
    [],
  );

  const [toggles, setToggles] = useState(['icon']);
  const handleToggleChange = useCallback((value) => setToggles(value), []);
  const isPropToggled = (value: string) => toggles.includes(value);

  return (
    <Page title="Button Configuration">
      <Layout>
        <Layout.Section oneThird>
          <Card title="Component">
            <Card.Section title="Result">
              <Button
                id="ButtonAllProps"
                ariaControls="ControlTarget"
                accessibilityLabel="Button a11y text"
                size={sizeOptions[size]}
                textAlign={textAlignOptions[textAlign]}
                disclosure={disclosureOptions[disclosure]}
                icon={isPropToggled('icon') ? OnlineStoreMajor : undefined}
                primary={isPropToggled('primary')}
                destructive={isPropToggled('destructive')}
                outline={isPropToggled('outline')}
                fullWidth={isPropToggled('fullWidth')}
                plain={isPropToggled('plain')}
                monochrome={isPropToggled('monochrome')}
                stretchContent={isPropToggled('stretchContent')}
                connectedDisclosure={
                  isPropToggled('connectedDisclosure')
                    ? mockConnectedDisclosure
                    : undefined
                }
                url={isPropToggled('url') ? mockUrl : undefined}
                external={isPropToggled('external')}
                download={isPropToggled('download')}
                submit={isPropToggled('submit')}
                disabled={isPropToggled('disabled')}
                loading={isPropToggled('loading')}
                pressed={isPropToggled('pressed')}
                role={isPropToggled('role') ? mockRole : undefined}
                ariaExpanded={isPropToggled('ariaExpanded')}
                onClick={mockEvent('onAction')}
                onFocus={mockEvent('onFocus')}
                onBlur={mockEvent('onBlur')}
                onKeyPress={mockEvent('onKeyPress')}
                onKeyUp={mockEvent('onKeyUp')}
                onKeyDown={mockEvent('onKeyDown')}
                onMouseEnter={mockEvent('onMouseEnter')}
                onTouchStart={mockEvent('onTouchStart')}
              >
                {childrenOptions[children]}
              </Button>
            </Card.Section>
            <Card.Section title="Output">
              <p id="ControlTarget">
                <strong>Boolean toggles:</strong> {toggles.join(', ')}
              </p>
            </Card.Section>
          </Card>
        </Layout.Section>

        <Layout.Section oneThird>
          <Card title="Children" sectioned>
            <ChoiceList
              title=""
              choices={Object.keys(childrenOptions).map((prop) => ({
                label: childrenOptions[prop] ?? 'None',
                value: prop,
              }))}
              selected={children}
              onChange={handleChildrenChange}
            />
          </Card>

          <Card title="Size" sectioned>
            <ChoiceList
              title=""
              choices={Object.keys(sizeOptions).map((prop) => ({
                label: sizeOptions[prop] ?? 'None',
                value: prop,
              }))}
              selected={size}
              onChange={handleSizeChange}
            />
          </Card>

          <Card title="Text alignment" sectioned>
            <ChoiceList
              title=""
              choices={Object.keys(textAlignOptions).map((prop) => ({
                label: textAlignOptions[prop] ?? 'None',
                value: prop,
              }))}
              selected={textAlign}
              onChange={handleTextAlignChange}
            />
          </Card>

          <Card title="Disclosure" sectioned>
            <ChoiceList
              title=""
              choices={Object.keys(disclosureOptions).map((prop) => ({
                label: disclosureOptions[prop] ?? 'None',
                value: prop,
              }))}
              selected={disclosure}
              onChange={handleDisclosureChange}
            />
          </Card>
        </Layout.Section>

        <Layout.Section oneThird>
          <Card title="Toggles" sectioned>
            <ChoiceList
              allowMultiple
              title=""
              choices={toggleOptions.map((prop) => ({
                label: prop,
                value: prop,
              }))}
              selected={toggles}
              onChange={handleToggleChange}
            />
          </Card>
        </Layout.Section>
      </Layout>
    </Page>
  );
}
```

</details>

**This PR makes a few small adjustments to the `Button` component:**
- ~Enable `color` transitions, allowing for `newDesignLanguage` "icons" to inherit `color` from their parent and transition accordingly.~
- Revise use of `transition`, as it was inconsistent.
  - Avoid `transition` for "old design language"
  - Enable `transition` for "new design language"
  - Remove use of `--p-override-none`
- Fix issue with `Button` text remaining visible when `newDesignLanguage + loading + plain`.
- No longer do a `180deg` spin on the `disclosure` when toggling between `up/down`.
- Right-align the `disclosure` arrow when working with `textAlignLeft`.
- No longer set `border-width: 0` on "filled" buttons, which was resulting in layout shift when toggling certain boolean props.
- A few minor code adjustments to improve readability.
- New `"select"` option for `disclosure`.

### Screenshots

<details>
<summary><strong>1. Revised transitions:</strong></summary>

transitions

</details>

<details>
<summary><strong>2. `newDesignLanguage + loading + plain`</strong></summary>

button-plain-loading

</details>

<details>
<summary><strong>3. No longer spin `disclosure`:</strong></summary>

disclosure-up-down

</details>

<details>
<summary><strong>4. Right-align `disclosure`:</strong></summary>

disclosure-right

</details>

<details>
<summary><strong>5. Prevent layout-shift from `border-width`:</strong></summary>

border-width-layout

</details>

<details>
<summary><strong>6. Select disclosure:</strong></summary>

Screen Shot 2020-12-15 at 4 06 28 PM

</details>

---

### Questions:

**1. Should we remove the left/right margins when "only disclosure"?**
Its a low-effort fix... but maybe we want to discourage "only disclosure"?

<details>
<summary><strong>Screenshot:</strong></summary>

</details>

**~2. Should `stretchContent` instead be a `textAlign="justified"` option?~**
~Another low-effort fix. Also... why do we need to qualify the styles with `:not(.loading)`?~
Since resolved. Marked for deprecation.

**3. Why were we using `--p-override-none`?**
This was used on `transition-duration` and `transition-timing-function`. Seems to suggest we want to disable transitions in "new design language" but enable for "old design language"? This PR reverses that behaviour while completely removing `--p-override-none` from `Button.scss`. Am I misunderstanding something? Should I revert?

**4. What is with the `.monochrome.outline > ::before` for?**
Seems like its used to provide a "background". So what I don't understand is: why not use `background-color`?

**5. Do we want to document/implement logic for "incompatible props"?**
Several prop combinations are incompatible, such as: `plain + connectedDisclosure` or `primary + destructive`.

**6. Should the deprecated `ariaPressed` not have been removed by now?**
Am I able to get rid of this prop from the `BaseButton` type?

**7. Should we swap `icon` and `disclosure` for `rtl` languages?**
My instinct is "yes". Have we received any feedback about this?

---

### Pre-existing issues (discovered while tophatting):

1. `loading` does not affect `connectedDisclosure`
2. Undesired styles when using: `outline + disabled + connectedDisclosure`
3. Undesired styles when using: `primary + outline`
4. There are surely a few more faulty prop combinations... but I ran out of time tophatting.

cc/ @emma-boardman in case you have some extra time to help tophat.
cc/ @dleroux since you have touched some of the changed code
cc/ @dfmcphee since you also have some expertise/opinions on this component

### CODE REVIEW

**Comment by BPScott** [comment]
Heya @beefchimi, you might want to check out https://github.com/Shopify/polaris-react/pull/3666#issuecomment-745690844

File: src/styles/shared/_buttons.scss
Comment (dleroux) [inline src/styles/shared/_buttons.scss:35]: What's the timeline for removing these? Won't that break apps using the component if they haven't enabled the new design language?

Comment (beefchimi) [reply src/styles/shared/_buttons.scss:35]: @dleroux was this ever working correctly though? There was a comment in the code about doing _something a certain way_ in order to animate the `background: linear-gradient`, but as far as I know, you can't animate gradients. So we had some properties animating while others were not, adding up to a disjointed experience.

Worth mentioning though - the properties that were animating correctly were things like `border-color` and `box-shadow`, and I think they were hard to notice in the first place. So the experience wasn't terribly inconsistent, unless you were _really paying attention_.

Essentially, I don't really see the drawback of doing this for old design language users. But I completely defer to Polaris here - you folks are going to have way better context than I will 😆

Comment (dleroux) [reply src/styles/shared/_buttons.scss:35]: Ha yes, It was a safari bug IIRC. @danrosenthal and @tmlayton were involved in that.

> you folks are going to have way better context than I will.

I do have context but I'm not on Polaris anymore :)

File: src/types.ts
Comment (beefchimi) [inline src/types.ts:97]: This was pointing to the wrong link. But also - as mentioned in my `Questions` section - should we not have removed this already?

File: src/styles/shared/_buttons.scss
Comment (danrosenthal) [reply src/styles/shared/_buttons.scss:35]: No issues related to the Safari bug here. That was related to the `linear-gradient` function being passed as a `var` function fallback property.

**Comment by beefchimi** [comment]
@BPScott / @dleroux this is ready for final review 👍🏻

**Comment by beefchimi** [comment]
@robertmeessen + @BPScott this should be ready for final review.

Is there someone that reviews the `percy/codecov` changes and approves them?

**Review by robertmeessen — APPROVED** [review]
🎩  Looks good to me, thanks @beefchimi

File: UNRELEASED.md
Comment (alex-page) [inline UNRELEASED.md:13]:
```suggestion
- Right-align `disclosure` when using `textAlignLeft` for `Button` ([#3709](https://github.com/Shopify/polaris-react/pull/3709))
- More consistent transitions in `Button` ([#3709](https://github.com/Shopify/polaris-react/pull/3709))
- New `select` option for `disclosure` in `Button` ([#3709](https://github.com/Shopify/polaris-react/pull/3709))
```

File: src/components/Button/Button.scss
Comment (alex-page) [inline src/components/Button/Button.scss:163]: Any reason for change here?

**Comment by BPScott** [comment]
Thanks for putting up with me. Gonna look at this later today :)

File: src/styles/shared/_buttons.scss
Comment (alex-page) [inline src/styles/shared/_buttons.scss:77]: Lets remove all transitions. This is not adding any benefit to the user experience right now.

File: UNRELEASED.md
Comment (beefchimi) [reply UNRELEASED.md:13]: Will make the change, but I do think its worth considering this style in the future. We write things this way in `online-store-ui` and I find it much easier to parse the Release Notes and identify the components that are relevant to the consuming app.

As a consumer, when I'm going through `Breaking changes`, I can quickly look at the prefixed components and identify what is relevant to my app.

File: src/styles/shared/_buttons.scss
Comment (alex-page) [reply src/styles/shared/_buttons.scss]: We will spend some time looking into component transitions in the future, for now lets remove them.

Comment (beefchimi) [reply src/styles/shared/_buttons.scss]: I agree that the `duration` was too slow. 👍🏻 

Also agree with the sentiment that for subtle changes, motion can make them _even more subtle_, therefor lessening the discoverability of the change. However, after speeding up the `duration`, I'm not sure that point is as relevant anymore?

If we are committed to completely removing the button transitions - I'd ask that we consider offering some CSS custom property values that consumers can override. That way, Polaris can default to no transition, but consumers can easily override.

If that is a path we want to pursue - and if there is a non-insignificant time investment - I would then ask if we can defer that to a follow-up PR? Getting this PR merged and available to `online-store-ui` would help us unblock ourselves.

File: src/components/Button/tests/Button.test.tsx
Comment (BPScott) [inline src/components/Button/tests/Button.test.tsx:451]: Perform mocking and reseting of mocks in a beforeEach / afterEach method rather than within the test block. Eg. https://github.com/Shopify/polaris-react/blob/3d2da225b7c484e446dd1522cab0db0dfbfdbb87/src/components/Modal/components/Dialog/tests/Dialog.test.tsx#L10-L16

This is because if the mockRestore is within the test block and an expectation fails then the mockResore() won't be called and that mock will potentially leak in other test cases, which can lead to hard to diagnose cascading errors - SomeOther.test.tsx failing because it is inadvertently using a mock that was set but not cleared up in Button.test.tsx is hecking confusing (I've been bit by this before, it's a great way to waste an afternoon)

File: src/types.ts
Comment (BPScott) [reply src/types.ts:99]: Good spot! Yeah this should have ideally been removed in v5.0.0 but we missed that window. We should try again in v6.0.0

**Review by BPScott — COMMENTED** [review]
Looking very tasty.

Something we talked about a while back was creating this layout:

Screen Shot 2021-01-06 at 1 15 37 PM

I can create that with:

```jsx
<Button
  icon={OnlineStoreMajor}
  disclosure="select"
  fullWidth
  textAlign="center"
  stretchContent
>
  Some Text
</Button>
```

But i need stretchContent push the disclosure and icon to the edges of the box.

Is this a style of layout we want to support? I don't think we can currently create that if we deprecate and remove stretchContent?

**Comment by BPScott** [comment]
The plot thickens: Things are stretched and the icon and disclosure are at the edges of their boxes out when textAlign is left, but not when right or centered:

Screen Shot 2021-01-06 at 1 27 22 PM

Produced by:

```
<Button
  icon={OnlineStoreMajor}
  disclosure="select"
  fullWidth
  textAlign="left"
>
  align=left
</Button>
<br />
<Button
  icon={OnlineStoreMajor}
  disclosure="select"
  fullWidth
  textAlign="center"
>
  align=center
</Button>
<br />
<Button
  icon={OnlineStoreMajor}
  disclosure="select"
  fullWidth
  textAlign="right"
>
  align=right
</Button>
```

Should we handle that? My initial gut feel is that this is weird, but sitting on it more this might actually be reasonable behaviour - if you get rid either the icon or the disclosure then things feel lopsided.

Perhaps the fix is is "the discource icon is always flush to the right hand edge" and "When you talk about alignment, you consider the icon text as a single block and you align that at the same time"?

**Comment by BPScott** [comment]
Another nitpicky thing:

When specifying textAlign="left" the padding of the disclosure icon changes. It has no effect when set to any other value. I'm thinking the following buttons should all render visually identically

Screen Shot 2021-01-06 at 1 45 50 PM

Produced by:

```
<Button disclosure="select">align</Button>
<br />
<Button disclosure="select">align</Button>
<br />
<Button disclosure="select" textAlign="left">
  align
</Button>
<br />
<Button disclosure="select" textAlign="center">
  align
</Button>
<br />
<Button disclosure="select" textAlign="right">
  align
</Button>
```

File: src/styles/shared/_buttons.scss
Comment (alex-page) [reply src/styles/shared/_buttons.scss]: I think this PR can remove the transitions, instead of making them consistent. Let me know if you want me to do the work ^

Comment (beefchimi) [reply src/styles/shared/_buttons.scss]: I can easily remove the transitions - I was hoping however we could compromise and go the route of CSS Custom Properties so that Polaris can default to no transitions, but I - a consumer - can more easily enable transitions without digging into underlying markup.

Removing transitions altogether invalidates one of my approaches to utilizing Polaris' `Button`:
https://github.com/Shopify/online-store-ui/pull/2143

To be honest I didn't love that solution anyways _(for various reasons)_, its a very gross way of customizing a component, but it did save me from copying much of `Button.tsx` over into my codebase.

Removing transitions forces me into the other approach:
https://github.com/Shopify/online-store-ui/pull/2199

At this point the PR isn't actually providing me any value. Debatable if we want to continue with pushing this forward?

Comment (alex-page) [reply src/styles/shared/_buttons.scss]: Why do the online-store buttons require transitions? Right now I am saying that by default all buttons should not have them?

**Comment by beefchimi** [comment]
@BPScott 

> Something we talked about a while back was creating this layout:

Oh I had thought that we didn't want to support that layout anymore? If we do want to support it, I suppose we could make this what happens when you use: `textAlign="center" + fullWidth + disclosure`.

> Things are stretched and the icon and disclosure are at the edges of their boxes out when textAlign is left, but not when right or centered:

This was also discussed during our meeting. I was under the impression we were happy with this behaviour, but I can change it. What you want to see is: `disclosure + fullWidth` ALWAYS results in a right-aligned `disclosure`?

I don't have super strong opinions here because I don't actually think all of these prop combinations are going to be recommended patterns. This kind of leads me back to my question about combo-prop validation and how we define logic around what props can be combo'd together. At the end of the day I think we should just let the consumer do what they want and not be so defensive, but the trade off there is "do we then have to spend time making every combination look good?".

> When specifying textAlign="left" the padding of the disclosure icon changes.

Ah yes good catch, that would be a [result of this commit](https://github.com/Shopify/polaris-react/pull/3709/commits/0f1f69be789ba696859975c04945c34ba16f8312) where we decided to no longer scope the `disclosure` alignment to `fullWidth`. I think this is now a valid case to go back to the added specificity.

> Perform mocking and reseting of mocks in a beforeEach / afterEach method

heheh... I was just copying this style of test authoring: https://github.com/Shopify/polaris-react/blob/master/src/components/Button/tests/Button.test.tsx#L175

I will fix them both! Since there are two of them, I did the mocking at the beginning of the test suite rather than in each `describe`. Did you prefer I do this twice in each `describe`?

As always, thanks for the info :) this will no doubt save me from trouble in the future.

### Closing thoughts

So I spent some time trying to accommodate all of your feedback... and it is very hard! No matter what I always end up with _some prop combination_ revealing an undesirable result. I think in order to do this correctly, I need to introduce a `.hasDisclosure` variant class. I feel like this is going down a dubious path. `Button` has ended up trying to do too many things and resulted in too many conditions to account for.

In order to really go through each scenario and craft the best possible combination, it will take me more time than I have available right now. So before attempting to solve everything, the questions I'd ask are:
1. Is the new result worse/same/better than before?
2. Is this a prop combination we would ever want to see in the wild?

File: src/styles/shared/_buttons.scss
Comment (beefchimi) [reply src/styles/shared/_buttons.scss]: > all buttons should not have them

I wouldn't have expected that opinion to be extended to consumers. I'm left wondering how we determine what styles are open to theme customization while others - such as transitions - are not? How can contributors better understand what fits into theme customization?

I think its fair to say that we have unique concerns for our app. While we conform to the system for much of what we do, we have our own set of problems to solve and therefor our own UI library to achieve that experience.

We utilize motion in much of our UI to enhance the experience. When you remove motion from some of our actions, it begins to feel out of place in respect to the rest of the UI.

I've misinterpreted systems-of-systems, as I wouldn't have expected us to be unable to utilize transitions for our actions. This feels like the wrong decision to me. What led me to this PR was an effort to remove a completely custom UI component in favour of consuming `Polaris.Button` and theming it. I think consumer's will be incentivized to keep building custom components so they can have better control over the desired styles.

When consumers write their own components, where do we look to inform ourselves about how to author certain UI details, such as motion? The currently styleguide doesn't really tell us very much about authoring our own motion styling, and it doesn't seem to include the context you provided from Johan:
https://polaris.shopify.com/design/polaris-is-changing#section-motion

Comment (alex-page) [reply src/styles/shared/_buttons.scss]: I think this is getting a bit into the weeds. We don't want consumers modifying the button transitions yet. We need to do exploration into how buttons should feel and animate.

File: UNRELEASED.md
Comment (alex-page) [reply UNRELEASED.md]: I like this format. I think we should be thinking about these sorts of things in a seperate PR or issue though.

**Comment by BPScott** [comment]
> Oh I had thought that we didn't want to support that layout anymore? If we do want to support it, I suppose we could make this what happens when you use: textAlign="center" + fullWidth + disclosure.

...

> This was also discussed during our meeting. I was under the impression we were happy with this behaviour, but I can change it. What you want to see is: disclosure + fullWidth ALWAYS results in a right-aligned disclosure?

My bad for forgetting outcomes of that meeting :)


> In order to really go through each scenario and craft the best possible combination, it will take me more time than I have available right now. So before attempting to solve everything, the questions I'd ask are:
> 
>Is the new result worse/same/better than before?
> Is this a prop combination we would ever want to see in the wild?


You're right, that this is indeed better than before I do want to be mindful of your time as getting lots of these combos to play nice is a time-consuming game of whack-a-mole. Let's call this good to merge now and we can revisit the center aligning at some point in the future if it becomes an issue again.

**Review by BPScott — APPROVED** [review]
Amazing work, thank you for picking through component full of edge cases and working out how to make improvements!

:shipit:

**Comment by alex-page** [comment]
Thank you @beefchimi 🚀

**Pull request merged**
